README.md
BestCaptchaSolver.com python2 API wrapper
=========================================

bestcaptchasolver-python2 is a super easy to use bypass captcha python2 API wrapper for bestcaptchasolver.com captcha service

## Installation    
    git clone https://github.com/bestcaptchasolver/bestcaptchasolver-python2

## Dependencies
    pip install requests

## Usage
    # make sure you've changed access_key, page_url, etc in main.py
    python main.py  

## How to use?

Simply require the module, set the auth details and start using the captcha service:

``` python
from bestcaptchasolverapi2 import BestCaptchaSolverAPI
```
Set access_token for authentication

``` python
access_token = 'access_token_here'
# get your access token from https://bestcaptchasolver.com/account
bcs = BestCaptchaSolverAPI(ACCESS_TOKEN)
```

Once you've set your authentication details, you can start using the API

**Get balance**

``` python
balance = bcs.account_balance()                 
```

**Solve image captcha**

``` python
data = {}
data['image'] = 'captcha.jpg'

# optional parameters
data['is_case'] = if case sensitive set to True, default: False
data['is_phrase'] = if phrase, set to True, default: False
data['is_math'] = True if captcha is math, default: False
data['alphanumeric'] = 1 (digits only) or 2 (letters only), default: all characters
data['minlength'] = minimum length of captcha text, default: any
data['maxlength'] = maximum length of captcha text, default: any

bcs.submit_image_captcha(data)
```
The image submission works with both files and b64 encoded strings.
For setting the affiliate_id, set the `affiliate_id` parameter

**Submit recaptcha details**

For recaptcha submission there are two things that are required.
- page_url
- site_key
- type (optional)
- v3_action (optional)
- v3_min_score (optional)
- user_agent (optional)
- affiliate_id (optional)
- proxy (optional)

``` python
captcha_id = bcs.submit_recaptcha({'page_url': 'page_url_here', 'site_key': 'sitekey_here')   
```

This method returns a captchaID. This ID will be used next, to retrieve the g-response, once workers have 
completed the captcha. This takes somewhere between 10-80 seconds.

**Geetest**
- domain
- gt
- challenge

```python
captcha_id = bcs.submit_geetest({'domain': 'DOMAIN_HERE', 'gt': 'GT_HERE', 'challenge': 'CHALLENGE_HERE'})
```

Use captcha_id to retrieve `solution` for geetest

**Capy**
- page_url
- site_key

```python
captcha_id = bcs.submit_capy({'page_url': 'PAGE_URL_HERE', 'site_key': 'SITEKEY_HERE'})
```

Use captcha_id to retrieve `solution` for capy

**Retrieve captcha response (all captchas)**

```
image_text = bcs.retrieve(captcha_id)['text']
gresponse = bcs.retrieve(recaptcha_id)['gresponse']
solution = bcs.retrieve(recaptcha_id)['solution']
```

**If submitted with proxy, get proxy status**
```
proxy_status = bcs.retrieve(recaptcha_id)['proxy_status']
```

**Set captcha bad**

When a captcha was solved wrong by our workers, you can notify the server with it's ID,
so we know something went wrong.

``` python
bcs.set_captcha_bad(captcha_id)
```

## Examples
Check main.py

## License
API library is licensed under the MIT License

## More information
More details about the server-side API can be found [here](https://bestcaptchasolver.com/api )


<sup><sub>captcha, bypasscaptcha, decaptcher, decaptcha, 2captcha, deathbycaptcha, anticaptcha, 
bypassrecaptchav2, bypassnocaptcharecaptcha, bypassinvisiblerecaptcha, captchaservicesforrecaptchav2, 
recaptchav2captchasolver, googlerecaptchasolver, recaptchasolverpython, recaptchabypassscript, bestcaptchasolver</sup></sub>



changelog.md
23.09.2019
----------
- added new image captcha optional parameters
- added geetest captcha
- added capy captcha
- splitted `main.py` into multiple example files

examples/capy.py
#!/usr/bin/python2.7

from bestcaptchasolverapi2 import BestCaptchaSolverAPI
from time import sleep

ACCESS_TOKEN = 'ACCESS_TOKEN_HERE'

# for more details check https://bestcaptchasolver.com/captchabypass-api
def test_api():
    bcs = BestCaptchaSolverAPI(ACCESS_TOKEN)        # get access token from: https://bestcaptchasolver.com/account

    # check account balance
    # ---------------------------
    balance = bcs.account_balance()                       # get account balance
    print ('Balance: {}'.format(balance))                 # print balance

    print ('Waiting for capy to be solved ...')
    solution = None
    captcha_id = bcs.submit_capy({'page_url': 'PAGE_URL_HERE', 'site_key': 'SITEKEY_HERE'})
    while solution == None:  # while it's still in progress
        resp = bcs.retrieve(captcha_id)
        solution = resp['solution']
        sleep(10)  # sleep for 10 seconds and recheck
    print (solution)

    # bcs.set_captcha_bad(2)    # set captcha with ID 2, bad

# main method
def main():
    try:
        test_api()
    except Exception, ex:
        print ('[!] Error occured: {}'.format(ex))

if __name__ == "__main__":
    main()


examples/image.py
#!/usr/bin/python2.7

from bestcaptchasolverapi2 import BestCaptchaSolverAPI
from time import sleep

ACCESS_TOKEN = 'ACCESS_TOKEN_HERE'

# for more details check https://bestcaptchasolver.com/captchabypass-api
def test_api():
    bcs = BestCaptchaSolverAPI(ACCESS_TOKEN)        # get access token from: https://bestcaptchasolver.com/account

    # check account balance
    # ---------------------------
    balance = bcs.account_balance()                       # get account balance
    print ('Balance: {}'.format(balance))                 # print balance

    print ('Solving image captcha ...')
    data = {}
    data['image'] = '../captcha.jpg'

    # optional parameters
    # -------------------
    # data['is_case'] = True, default: False
    # data['is_phrase'] = True, default: False
    # data['is_math'] = True, default: False
    # data['alphanumeric'] = 1 (digits only) or 2 (letters only), default: all characters
    # data['minlength'] = minimum length of captcha text, default: any
    # data['maxlength'] = maximum length of captcha text, default: any
    # data['affiliate_id'] = 'affiliate_id from /account'

    id = bcs.submit_image_captcha(data)  # submit image captcha (case_sensitive param optional)
    image_text = None
    # None is returned if completion is still in pending
    while image_text == None:
        image_text = bcs.retrieve(id)['text']  # get the image text using the ID
        sleep(5)

    print ('Captcha text: {}'.format(image_text))

# main method
def main():
    try:
        test_api()
    except Exception, ex:
        print ('[!] Error occured: {}'.format(ex))

if __name__ == "__main__":
    main()


examples/geetest.py
#!/usr/bin/python2.7

from bestcaptchasolverapi2 import BestCaptchaSolverAPI
from time import sleep

ACCESS_TOKEN = 'ACCESS_TOKEN_HERE'

# for more details check https://bestcaptchasolver.com/captchabypass-api
def test_api():
    bcs = BestCaptchaSolverAPI(ACCESS_TOKEN)        # get access token from: https://bestcaptchasolver.com/account

    # check account balance
    # ---------------------------
    balance = bcs.account_balance()                       # get account balance
    print ('Balance: {}'.format(balance))                 # print balance

    solution = None
    captcha_id = bcs.submit_geetest({'domain': 'DOMAIN_HERE', 'gt': 'GT_HERE', 'challenge': 'CHALLENGE_HERE'})
    print ('Waiting for geetest to be solved ...')
    while solution == None:    # while it's still in progress
        resp = bcs.retrieve(captcha_id)
        solution = resp['solution']
        sleep(10)               # sleep for 10 seconds and recheck
    print (solution)

    # bcs.set_captcha_bad(2)    # set captcha with ID 2, bad

# main method
def main():
    try:
        test_api()
    except Exception, ex:
        print ('[!] Error occured: {}'.format(ex))

if __name__ == "__main__":
    main()


examples/recaptcha.py
#!/usr/bin/python2.7

from bestcaptchasolverapi2 import BestCaptchaSolverAPI
from time import sleep

ACCESS_TOKEN = 'ACCESS_TOKEN_HERE'
PAGE_URL = 'PAGE_URL_HERE'
SITE_KEY = 'SITE_KEY_HERE'

# for more details check https://bestcaptchasolver.com/captchabypass-api
def test_api():
    bcs = BestCaptchaSolverAPI(ACCESS_TOKEN)        # get access token from: https://bestcaptchasolver.com/account

    # check account balance
    # ---------------------------
    balance = bcs.account_balance()                       # get account balance
    print ('Balance: {}'.format(balance))                 # print balance

    # solve recaptcha
    # ---------------
    print ('Solving recaptcha ...')
    data = {}
    data['page_url'] = PAGE_URL
    data['site_key'] = SITE_KEY

    # other parameters
    # ----------------------------------------------------------------------
    #data['type'] = 1        # 1 - regular, 2 - invisible, 3 - v3, default 1
    #data['v3_action'] = 'v3 recaptcha action'
    #data['v3_min_score'] = '0.3'
    #data['user_agent'] = 'Your user agent'
    #data['proxy'] = '123.456.678:3031'
    #data['proxy'] = 'user:pass@123.456.678:3031'
    #data['affiliate_id'] = 'affiliate_id from /account'
    captcha_id = bcs.submit_recaptcha(data)        # submit captcha first, to get ID

    # check if it's still in progress (waiting to be solved), every 10 seconds
    print ('Waiting for recaptcha to be solved ...')
    gresponse = None
    while gresponse == None:    # while it's still in progress
        resp = bcs.retrieve(captcha_id)
        gresponse = resp['gresponse']
        sleep(10)               # sleep for 10 seconds and recheck

    print ('Recaptcha response: {}'.format(gresponse))         # print google response

    # proxy_status = resp['proxy_status']                      # get status of proxy
    # bcs.set_captcha_bad(2)    # set captcha with ID 2, bad

# main method
def main():
    try:
        test_api()
    except Exception, ex:
        print ('[!] Error occured: {}'.format(ex))

if __name__ == "__main__":
    main()


bestcaptchasolverapi2/bestcaptchasolverapi.py
try:
    from requests import session
    import requests
except:
    raise Exception('requests package not installed, try with: \'pip2.7 install requests\'')

import os, json
from base64 import b64encode

USER_AGENT = 'python2Client'
BASE_URL = 'https://bcsapi.xyz/api'
SSL_VERIFY = True
# endpoints
# -------------------------------------------------------------------------------------------


# API class
# -----------------------------------------
class BestCaptchaSolverAPI:
    def __init__(self, access_token, timeout = 120):
        self._access_token = access_token
        self._data = {
            'access_token': access_token
        }

        self._timeout = timeout
        self._session = session()       # init a new session

        self._headers = {               # use this user agent
            'User-Agent' : USER_AGENT
        }

    # get account balance
    def account_balance(self):
        url = '{}/user/balance?access_token={}'.format(BASE_URL, self._access_token)
        resp = self.GET(url)
        return '${}'.format(resp['balance'])

    # solve normal captcha
    def submit_image_captcha(self, opts):
        data = {}
        data.update(self._data)

        image_path = opts['image']
        # optional parameters
        if opts.has_key('case_sensitive'):
            print ('case_sensitive is deprecated, use is_case instead')
            if opts['case_sensitive']: data['is_case'] = True
        if opts.has_key('is_case'):
            if opts['is_case']: data['is_case'] = True
        if opts.has_key('is_phrase'):
            if opts['is_phrase']: data['is_phrase'] = True
        if opts.has_key('is_math'):
            if opts['is_math']: data['is_math'] = True
        if opts.has_key('alphanumeric'): data['alphanumeric'] = opts['alphanumeric']
        if opts.has_key('minlength'): data['minlength'] = opts['minlength']
        if opts.has_key('maxlength'): data['maxlength'] = opts['maxlength']

        # affiliate
        if opts.has_key('affiliate_id'):
            if opts['affiliate_id']: data['affiliate_id'] = opts['affiliate_id']
        url = '{}/captcha/image'.format(BASE_URL)
        if os.path.exists(image_path):
            with open(image_path, 'r') as f:
                data['b64image'] = b64encode(f.read())
        else:
            data['b64image'] = image_path       # should be b64 already

        resp = self.POST(url, data)
        return resp['id']       # return ID

    # submit recaptcha to system
    def submit_recaptcha(self, data):
        data.update(self._data)
        if data.has_key('proxy'): data['proxy_type'] = 'HTTP' # add proxy, if necessary
        # make request with all data
        url = '{}/captcha/recaptcha'.format(BASE_URL)
        resp = self.POST(url, data)
        return resp['id']  # return ID

    # submit recaptcha to system
    def submit_geetest(self, data):
        data.update(self._data)
        if data.has_key('proxy'): data['proxy_type'] = 'HTTP' # add proxy, if necessary
        # make request with all data
        url = '{}/captcha/geetest'.format(BASE_URL)
        resp = self.POST(url, data)
        return resp['id']  # return ID

    # submit recaptcha to system
    def submit_capy(self, data):
        data.update(self._data)
        if data.has_key('proxy'): data['proxy_type'] = 'HTTP' # add proxy, if necessary
        # make request with all data
        url = '{}/captcha/capy'.format(BASE_URL)
        resp = self.POST(url, data)
        return resp['id']  # return ID

    # retrieve captcha
    def retrieve(self, captcha_id = None):
        url = '{}/captcha/{}?access_token={}'.format(BASE_URL, captcha_id, self._access_token)
        resp = self.GET(url)
        try:
            if resp['status'] == 'pending': return {'text': None, 'gresponse': None, 'solution': None}
        except:
            pass

        return resp
    # set captcha bad, if given id, otherwise set the last one
    def set_captcha_bad(self, captcha_id):
        data = dict(self._data)
        url = '{}/captcha/bad/{}'.format(BASE_URL, captcha_id)
        resp = self.POST(url, data)
        return resp['status']

    def GET(self, url):
        r = self._session.get(url, headers=self._headers, timeout=self._timeout, verify=SSL_VERIFY)
        js = json.loads(r.text.encode('utf-8'))
        if js['status'] == 'error': raise Exception(js['error'])
        return js

    def POST(self, url, data):
        r = self._session.post(url, data=data, headers=self._headers, timeout=self._timeout, verify=SSL_VERIFY)
        js = json.loads(r.text.encode('utf-8'))
        if js['status'] == 'error': raise Exception(js['error'])
        return js


main.py
#!/usr/bin/python2.7

from bestcaptchasolverapi2 import BestCaptchaSolverAPI
from time import sleep

ACCESS_TOKEN = 'your_access_token'
PAGE_URL = 'page_url_here'
SITE_KEY = 'sitekey_here'

# solve captcha
def test_api():
    bcs = BestCaptchaSolverAPI(ACCESS_TOKEN)        # get access token from: https://bestcaptchasolver.com/account

    # check account balance
    # ---------------------------
    balance = bcs.account_balance()                       # get account balance
    print ('Balance: {}'.format(balance))                 # print balance

    # solve image captcha
    # --------------------
    print ('Solving image captcha ...')
    data = {}
    data['image'] = 'captcha.jpg'
    #data['case_sensitive'] = True #, default: False
    #data['affiliate_id'] = 'affiliate_id from /account'
    id = bcs.submit_image_captcha(data)  # submit image captcha (case_sensitive param optional)
    image_text = None
    # None is returned if completion is still in pending
    while image_text == None:
        image_text = bcs.retrieve(id)['text']  # get the image text using the ID
        sleep(5)

    print ('Captcha text: {}'.format(image_text))

    # solve recaptcha
    # ---------------
    print ('Solving recaptcha ...')
    data = {}
    data['page_url'] = PAGE_URL
    data['site_key'] = SITE_KEY

    # other parameters
    # ----------------------------------------------------------------------
    #data['type'] = 1        # 1 - regular, 2 - invisible, 3 - v3, default 1
    #data['v3_action'] = 'v3 recaptcha action'
    #data['v3_min_score'] = '0.3'
    #data['user_agent'] = 'Your user agent'
    #data['proxy'] = '123.456.678:3031'
    #data['proxy'] = 'user:pass@123.456.678:3031'
    #data['affiliate_id'] = 'affiliate_id from /account'
    captcha_id = bcs.submit_recaptcha(data)        # submit captcha first, to get ID

    # check if it's still in progress (waiting to be solved), every 10 seconds
    print ('Waiting for recaptcha to be solved ...')
    gresponse = None
    while gresponse == None:    # while it's still in progress
        resp = bcs.retrieve(captcha_id)
        gresponse = resp['gresponse']
        sleep(10)               # sleep for 10 seconds and recheck

    print ('Recaptcha response: {}'.format(gresponse))         # print google response
    #proxy_status = resp['proxy_status']                       # get status of proxy
    # bcs.set_captcha_bad(2)    # set captcha with ID 2, bad

# main method
def main():
    try:
        test_api()
    except Exception, ex:
        print ('[!] Error occured: {}'.format(ex))

if __name__ == "__main__":
    main()
